### PR TITLE
[CARBONDATA-2644][DataLoad]ADD carbon.load.sortMemory.spill.percentage parameter  invalid value check

### PR DESCRIPTION
**Pull request by ndwangsen**
Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed?
 NA
 - [x] Any backward compatibility impacted?
 NO
 - [x] Document update required?
NO
 - [x] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        Testing done UT 
        - How it is tested? Please attach test report.
        add example for it
        - Is it a performance related change? Please attach the performance test report.
          NA
        - Any additional information to help reviewers in testing this change.
        NA
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
       NO

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Can one of the admins verify this patch?

**Comment by CarbonDataQA** [comment]
Can one of the admins verify this patch?

**Comment by ravipesala** [comment]
Can one of the admins verify this patch?

**Comment by chenliang613** [comment]
add to whitelist

**Comment by chenliang613** [comment]
Can you provide more background? why need to add this parameter?

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6467/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5298/

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonLoadOptionConstants.java [hunk 1]
```diff
@@ -144,8 +144,8 @@
    * If the sort memory is insufficient, spill inmemory pages to disk.
    * The total amount of pages is at most the specified percentage of total sort memory. Default
    * value 0 means that no pages will be spilled and the newly incoming pages will be spilled,
-   * whereas value 1 means that all pages will be spilled and newly incoming pages will be loaded
-   * into sort memory.
+   * whereas value 100 means that all pages will be spilled and newly incoming pages will be loaded
+   * into sort memory,Other percentage values range 0-100.
```
Comment (zzcclp) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonLoadOptionConstants.java]: add a space before 'Other'

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/TestLoadDataWithUnsafeMemory.scala [hunk 2]
```diff
@@ -64,6 +67,8 @@ class TestLoadDataWithUnsafeMemory extends QueryTest
       .addProperty(CarbonCommonConstants.UNSAFE_WORKING_MEMORY_IN_MB, "512")
     CarbonProperties.getInstance()
       .addProperty(CarbonCommonConstants.OFFHEAP_SORT_CHUNK_SIZE_IN_MB, "512")
+    CarbonProperties.getInstance()
+      .addProperty(CarbonLoadOptionConstants.CARBON_LOAD_SORT_MEMORY_SPILL_PERCENTAGE, "-1")
```
Comment (zzcclp) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/TestLoadDataWithUnsafeMemory.scala]: why use "-1" here?

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5389/

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonLoadOptionConstants.java (hunk 1 above)
Comment (ndwangsen) [reply core/src/main/java/org/apache/carbondata/core/constants/CarbonLoadOptionConstants.java]: ok

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/TestLoadDataWithUnsafeMemory.scala (hunk 2 above)
Comment (ndwangsen) [reply integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/TestLoadDataWithUnsafeMemory.scala]: test the range of 0-100

File: processing/src/main/java/org/apache/carbondata/processing/loading/sort/unsafe/merger/UnsafeIntermediateMerger.java [hunk 3]
```diff
@@ -88,6 +88,10 @@ public UnsafeIntermediateMerger(SortParameters parameters) {
           CarbonLoadOptionConstants.CARBON_LOAD_SORT_MEMORY_SPILL_PERCENTAGE,
           CarbonLoadOptionConstants.CARBON_LOAD_SORT_MEMORY_SPILL_PERCENTAGE_DEFAULT);
       spillPercentage = Integer.valueOf(spillPercentageStr);
+      if (spillPercentage > 100 || spillPercentage < 0) {
```
Comment (zzcclp) [inline processing/src/main/java/org/apache/carbondata/processing/loading/sort/unsafe/merger/UnsafeIntermediateMerger.java]: One suggestion: add a function, called 'getSortMemorySpillPercentage', in 'CarbonProperties' to get the value of 'CARBON_LOAD_SORT_MEMORY_SPILL_PERCENTAGE', and validate the value in this function. We don't need to validate this value again if other place need to use this config, just need to call 'getSortMemorySpillPercentage'.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6517/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5349/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5435/

File: processing/src/main/java/org/apache/carbondata/processing/loading/sort/unsafe/merger/UnsafeIntermediateMerger.java (hunk 3 above)
Comment (jackylk) [reply processing/src/main/java/org/apache/carbondata/processing/loading/sort/unsafe/merger/UnsafeIntermediateMerger.java]: property validation can be validated in `CarbonProperties.validateAndLoadDefaultProperties`, please add it there. And I agree with @zzcclp , you can add a function in `CarbonProperties` to get the value of this property, just like `getInsertIntoDatasetStorageLevel`

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonLoadOptionConstants.java [hunk 4]
```diff
@@ -144,8 +144,8 @@
    * If the sort memory is insufficient, spill inmemory pages to disk.
    * The total amount of pages is at most the specified percentage of total sort memory. Default
    * value 0 means that no pages will be spilled and the newly incoming pages will be spilled,
-   * whereas value 1 means that all pages will be spilled and newly incoming pages will be loaded
-   * into sort memory.
+   * whereas value 100 means that all pages will be spilled and newly incoming pages will be loaded
+   * into sort memory, Other percentage values range 0-100.
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonLoadOptionConstants.java]: please refine this sentence

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6646/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6648/

**Comment by ndwangsen** [comment]
retest this case please

**Comment by jackylk** [comment]
retest this please

**Comment by ndwangsen** [comment]
@chenliang613 This parameter controls how large sort temp file merge in memory

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6682/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5508/

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**